### PR TITLE
Resource merge

### PR DESCRIPTION
**Pull request by kdimopulu**

### CODE REVIEW

File: staff_features/resources/resource_merge.feature [hunk 1]
```diff
@@ -1,35 +1,34 @@
...
+Feature: Resource merge
   Background:
     Given an administrator user is logged in
-      And a Resource with name 'A' has been created
-      And a Resource with name 'B' has been created
-  Scenario: Merge two Resources by browsing Resource
-    Given the Resource 'A' is opened in edit mode
+      And two Resources have been created
+  Scenario: Merge two Resources by browsing
+    Given the first Resource is opened in edit mode
```
Comment (kdimopulu) [inline staff_features/resources/resource_merge.feature:6]: I think it is better to use Resource A and Resource B as in this way the steps are simpler to follow and more direct

File: staff_features/resources/resource_merge.feature [hunk 2]
```diff
@@ -1,35 +1,34 @@
...
-      And the user clicks on 'Link'
-      And the user clicks on 'Merge' in the confirm popup
-     Then the 'Resource' merged message is displayed
-      And the data from Resource 'B' is added to the Resource 'A'
-      And the Resource 'B' is deleted
-  Scenario: Merge two Resources by searching the Resource
-    Given the Resource 'A' is opened in edit mode
+      And the user clicks on the dropdown in the merge dropdown form
+      And the user clicks on 'Browse' in the merge dropdown form
+      And the user filters by text with the second Resource title in the modal
```
Comment (kdimopulu) [inline staff_features/resources/resource_merge.feature:10]: And the user filters by text with the 'B' Resource title in the modal

File: staff_features/resources/resource_merge.feature [hunk 3]
```diff
@@ -1,35 +1,34 @@
-Feature: Resource Merge
+Feature: Resource merge
   Background:
     Given an administrator user is logged in
-      And a Resource with name 'A' has been created
-      And a Resource with name 'B' has been created
-  Scenario: Merge two Resources by browsing Resource
-    Given the Resource 'A' is opened in edit mode
+      And two Resources have been created
```
Comment (kdimopulu) [inline staff_features/resources/resource_merge.feature:4]: And two Resources have been created wth names 'A' and 'B'

File: staff_features/resources/resource_merge.feature [hunk 4]
```diff
@@ -1,35 +1,34 @@
...
-      And the user clicks on 'Merge' in the confirm popup
-     Then the 'Resource' merged message is displayed
-      And the data from Resource 'B' is added to the Resource 'A'
-      And the Resource 'B' is deleted
-  Scenario: Merge two Resources by searching the Resource
-    Given the Resource 'A' is opened in edit mode
+      And the user clicks on the dropdown in the merge dropdown form
+      And the user clicks on 'Browse' in the merge dropdown form
+      And the user filters by text with the second Resource title in the modal
+      And the user selects the second Resource from the search results in the modal
```
Comment (kdimopulu) [inline staff_features/resources/resource_merge.feature:11]: And the user selects Resource 'B' from the search results in the modal

File: staff_features/resources/resource_merge.feature [hunk 5]
```diff
@@ -1,35 +1,34 @@
...
+      And the user clicks on 'Merge' in the modal
+     Then the 'Resource(s)' merged message is displayed
+      And the second Resource is deleted
+      And the following linked records from the second Resource are appended to the first Resource
+        | Related Accessions |
+        | Agent Links        |
+        | Subjects           |
+        | Classifications    |
+  Scenario: Merge two Resources by searching
+    Given the first Resource is opened in edit mode
```
Comment (kdimopulu) [inline staff_features/resources/resource_merge.feature:23]: Given the Resource 'A' is opened in edit mode

File: staff_features/resources/resource_merge.feature [hunk 6]
```diff
@@ -1,35 +1,34 @@
...
-       When the user clicks on 'Merge'
-        And the user fills in 'Resource' with 'B'
-        And the user selects 'B' from the suggestions
-        And the user clicks on 'Merge'
-        And the user clicks on 'Cancel' in the confirm popup
-       Then the confirm popup is disappeared
-        And the Resources are not changed
-      
\ No newline at end of file
+      And the user fills in and selects the second Resource in the merge dropdown form
```
Comment (kdimopulu) [inline staff_features/resources/resource_merge.feature:25]: And the user fills in and selects the 'B' Resource in the merge dropdown form

File: staff_features/resources/resource_merge.feature [hunk 7]
```diff
@@ -1,35 +1,34 @@
...
+      And the user clicks on 'Merge' in the merge dropdown form
+      And the user clicks on 'Merge' in the modal
+     Then the 'Resource(s)' merged message is displayed
+      And the second Resource is deleted
```
Comment (kdimopulu) [inline staff_features/resources/resource_merge.feature:29]: And the Resource 'B' is deleted

File: staff_features/resources/resource_merge.feature [hunk 8]
```diff
@@ -1,35 +1,34 @@
...
+      And the following linked records from the second Resource are appended to the first Resource
+        | Related Accessions |
+        | Agent Links        |
+        | Subjects           |
+        | Classifications    |
```
Comment (kdimopulu) [inline staff_features/resources/resource_merge.feature:34]: I would also add the Cancel Merge before confirmation scenario as this is testing an edge case and improves feature robustness.

Scenario: Cancel Merge Before Confirmation
Given the Resource 'A' is opened in edit mode
When the user clicks on 'Merge'
And the user fills in and selects the 'B' Resource in the merge dropdown form
And the user clicks on 'Merge' in the merge dropdown form
And the user clicks on 'Cancel' in the modal
Then the confirmation modal is dismissed
And Resource 'A' and Resource 'B' remain unchanged

**Comment by blacksmith-welder** [comment]
@kdimopulu 
I have replaced the words `first` and `second`, with `A` and `B` to reference the resources.
In my opinion, I would prefer the words `first` & `second` as they are more natural and accompany the phrase `two resources have been created`, even though `A` & `B` can work just fine.

However, the words `A` & `B` should be unquoted. We need concrete and separate steps here for the merge scenarios.
We use strings only for labels, buttons, user input fields, form titles, etc., for steps that we need them to be re-usable, and are subject to arbitrary user input.

Furthermore, the phrase `And two Resources have been created with names ... ` does not apply to resources, since a `Resource` does not have a `name` field, and it is quite ambiguous.
Resource has a title, so if we want to be compliant with the rest of the system, this should be something like:
`a resource with title '{string}' has been created`
which is not possible to do with the current configuration, since we need a unique title within the entire system, to avoid conflicts and failing tests.

**Review by brianzelip — COMMENTED** [review]
This PR looks good beyond the failing test which is likely not related to the PR.

File: staff_features/resources/step_definitions/resource_merge.rb [hunk 9]
```diff
@@ -0,0 +1,179 @@
+# frozen_string_literal: true
+
+Given 'two Resources A & B have been created' do
```
Comment (balmas) [inline staff_features/resources/step_definitions/resource_merge.rb:3]: I wonder if it would be better to have a helper function for creating resources. this code appears in number of places, and if anything about this process changes a lot of tests will need to be updated.

File: staff_features/resources/step_definitions/resource_shared.rb [hunk 10]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+When 'the user filters by text with the Resource title' do
```
Comment (balmas) [inline staff_features/resources/step_definitions/resource_shared.rb:3]: it doesn't look like this is called at least not in this PR - there's a separate "When 'the user filters by text with the Resource B title in the modal' do" in resource_merge.rb which looks very similar

File: staff_features/shared/step_definitions.rb [hunk 11]
```diff
@@ -83,6 +84,12 @@
   end
 end
 
+When 'the user clicks on {string} in the confirm modal' do |string|
```
Comment (balmas) [inline staff_features/shared/step_definitions.rb:87]: Maybe just an artifact of too many PRs queued, but this function (or one very similar) was removed in PR #38 and its being added back here, but doesn't seem to be called in this PR

Comment (blacksmith-welder) [reply staff_features/shared/step_definitions.rb]: Yes it has been removed somewhere on previous requests, but not from here. Removal was going to be applied on merge/rebase.

File: staff_features/resources/step_definitions/resource_shared.rb (hunk 10 above)
Comment (blacksmith-welder) [reply staff_features/resources/step_definitions/resource_shared.rb:3]: Yes it not required. It is going to be needed eventually, but it is removed for now.

File: staff_features/resources/step_definitions/resource_merge.rb (hunk 9 above)
Comment (blacksmith-welder) [reply staff_features/resources/step_definitions/resource_merge.rb:3]: I have thought of this since the beginning.
A step like `a Resource has been created`, could be considered a function and it's intended to be re-usable anywhere a resource is required.
This specific step (`two Resources A & B have been created`), is an edge case, in which we require two resources and it is currently intended for use only in this feature.
In the future, if we need multiple resources with differing underlying values, maybe we can create an abstraction for creating records.
However, the concern of future changes applies everywhere on these tests (and any tests in general). For example, here we are using Capybara, and we check for HTML elements in all step definitions.
